### PR TITLE
Feat - Voting Power

### PR DESCRIPTION
**Pull request by akyo8**

### CODE REVIEW

File: contracts/CreamVotingPower.sol [hunk 1]
```diff
@@ -1,12 +1,182 @@
...
+    address public constant sushiswap =
+        0xf169CeA51EB51774cF107c88309717ddA20be167;
+    address public constant masterchef =
+        0xc2EdaD668740f1aA35E4D8f227fB8E17dcA888Cd;
+    address public constant uniswap =
+        0xddF9b7a31b32EBAF5c064C80900046C9e5b7C65F;
+    address public constant balancer =
+        0x280267901C175565C64ACBD9A3c8F60705A72639;
+
+    uint256 public MINIMUM_VOTING_POWER = 1;
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]: should be 1e18

File: contracts/CreamVotingPower.sol [hunk 2]
```diff
@@ -1,12 +1,182 @@
...
+
+    function getVotingPower(address _holder) public view returns (uint256) {
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]: We'll like to conform the standard ERC20 interface for this function signature
`balanceOf`

File: scripts/sample-script.js [hunk 3]
```diff
@@ -1,32 +1,43 @@
...
 
-  await greeter.deployed();
+  await VotingPower.deployed();
+  console.log('VotingPower deployed to:', VotingPower.address);
 
-  console.log("Greeter deployed to:", greeter.address);
+  for (let i = 0; i < addresses.length; i++) {
+    try {
+      const response = await VotingPower.getVotingPower(addresses[i]);
+      console.log(`=========${addresses[i]}=========\n${response}`);
```
Comment (jeremyHD) [inline scripts/sample-script.js:29]: Oh nice, you have written the test that I was trying to tell you about! Sorry I didn't see this just now

File: scripts/example.json [hunk 4]
```diff
@@ -0,0 +1,9 @@
+[
+  "0xeB8A4eADd643Ea3873dfecFBa18BD709D3206919",
+  "0xe0Ba6c4c82376c76386D1Ca2ae2516Ab18882e3C",
+  "0xE32F576578F9aea942226B6A280fbCe93da45638",
+  "0x56e0150b3cbb9f191fe360740fcfa2edbd1b3697",
+  "0xbbe21151c726F9EEc42516524f58D0057261CC8a",
+  "0x6bD48938B9F9Dd02ee3D6dccd0379DDF0af22D38",
+  "0xa805cb6907130c0c7F3534Ee2EBc62D806FFB820"
```
Comment (jeremyHD) [inline scripts/example.json:8]: And you also bring the same addresses from the examples. Good Job!!!

If you set the `periods` in [example.json](https://github.com/snapshot-labs/snapshot.js/blob/master/src/strategies/cream/examples.json#L15) to `0` for snapshot strategy.
And run `npm run test --strategy=cream` on snapshot.js repo, the result should be the same or close enough

File: contracts/CreamVotingPower.sol [hunk 5]
```diff
@@ -1,12 +1,182 @@
...
+     @param _holder Address of holder
+     @return uint256 The cream token amount
+    */
+    function _lendingSupply(address _holder) internal view returns (uint256) {
+        require(
+            _holder != address(0),
+            "VotingPower.lendingSupply: Zero Address"
+        );
+
+        uint256 totalLending = ICToken(crCream).balanceOf(_holder) * ICToken(crCream).exchangeRateStored() / 1e18;
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]:
```suggestion
        uint256 totalLending = ICToken(crCream).balanceOf(_holder) * ICToken(crCream).exchangeRateStored().div( 1e18);
```

File: contracts/CreamVotingPower.sol [hunk 6]
```diff
@@ -1,12 +1,182 @@
...
+                .add(_sushiswapBalance(_holder))
+                .add(_uniswapBalance(_holder))
+                .add(_balancerBalance(_holder))
+                .add(_stakedInLPool(_holder))
+                .sub(_borrowedBalance(_holder));
+
+        require(
+            votingPower > MINIMUM_VOTING_POWER,
+            "VotingPower.getVotingPower: Invalid"
+        );
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]: return 0 if votingPower < Min voting power

File: scripts/example.json (hunk 4 above)
Comment (akyo8) [reply scripts/example.json:8]: @jeremyHD 
When `periods` was set `0`, it ran into this issue `Error: division by zero`.
Please tell me if you have any idea with it. Thanks

Comment (jeremyHD) [reply scripts/example.json:8]: set to 1. My bad

**Comment by akyo8** [comment]
@jeremyHD 
I updated with fixing of above comments and added defiPulse amount.
If you have any improvements or suggestions, please let me know. thanks

File: contracts/CreamVotingPower.sol [hunk 7]
```diff
@@ -1,12 +1,202 @@
...
 
-  function balanceOf(address account) returns (uint256) {
-    return 0;
-  }
+    struct UserInfo {
+        uint256 amount;
+        uint256 rewardDebt;
+    }
+
+    address public constant defiPulse = 0x1494CA1F11D487c2bBe4543E90080AeBa4BA3C2b;
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]: should be defiPulseIndex or just name it dpi

File: contracts/CreamVotingPower.sol [hunk 8]
```diff
@@ -1,12 +1,202 @@
...
+    */
+    function _lendingSupply(address _holder) internal view returns (uint256) {
+        require(
+            _holder != address(0),
+            "VotingPower.lendingSupply: Zero Address"
+        );
+
+        uint256 totalLending = ICToken(crCream).balanceOf(_holder).mul(ICToken(crCream).exchangeRateStored()).div(1e18);
+        uint256 borrowed = ICToken(crCream).borrowBalanceStored(_holder);
+
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol:97]: Please add a sanity check here
```
if (borrowed > totalLending) {
    return 0
}
```

or else safeMath will revert if underflow

File: contracts/CreamVotingPower.sol [hunk 9]
```diff
@@ -1,12 +1,202 @@
...
+    function _stakedInLPool(address _holder) internal view returns (uint256) {
+        require(
+            _holder != address(0),
+            "VotingPower.lendingSupply: Zero Address"
+        );
+
+        uint256 totalStaked = 0;
+
+        for (uint256 i = 0; i < 4; i++) {
+            totalStaked.add(ILPool(lPool[i]).balanceOf(_holder));
```
Comment (jeremyHD) [inline contracts/CreamVotingPower.sol]:
```suggestion
            totalStaked = totalStaked.add(ILPool(lPool[i]).balanceOf(_holder));
```

**Review by jeremyHD — CHANGES_REQUESTED** [review]

File: contracts/CreamVotingPower.sol (hunk 9 above)
Comment (akyo8) [reply contracts/CreamVotingPower.sol]: fixed

File: contracts/CreamVotingPower.sol (hunk 8 above)
Comment (akyo8) [reply contracts/CreamVotingPower.sol:97]: added sanity check for it

File: contracts/CreamVotingPower.sol (hunk 7 above)
Comment (akyo8) [reply contracts/CreamVotingPower.sol]: it's been named defiPulseIndex

**Comment by akyo8** [comment]
@jeremyHD 
Please check above fixing and this is final results of testing

=========0xeB8A4eADd643Ea3873dfecFBa18BD709D3206919=========
25988067596870823513528
=========0xe0Ba6c4c82376c76386D1Ca2ae2516Ab18882e3C=========
443299961957519273343
=========0xE32F576578F9aea942226B6A280fbCe93da45638=========
0
=========0x56e0150b3cbb9f191fe360740fcfa2edbd1b3697=========
83128575877236447792
=========0xbbe21151c726F9EEc42516524f58D0057261CC8a=========
0
=========0x6bD48938B9F9Dd02ee3D6dccd0379DDF0af22D38=========
51222026950618754185
=========0xa805cb6907130c0c7F3534Ee2EBc62D806FFB820=========
0

**Review by jeremyHD — APPROVED** [review]

**Pull request merged**